# Homework Assignment - Lesson 7: String Manipulation and Date/Time Data

**Student Name:** Charles Moten-Lopez

**Student ID:** @02126789     gan744

**Date Submitted:** October 19th, 2025

**Due Date:** October 12th, 2025

---

## Objective

Master string manipulation with `stringr` and date/time operations with `lubridate` for real-world business data cleaning and analysis.

## Learning Goals

By completing this assignment, you will:
- Clean and standardize messy text data using `stringr` functions
- Parse and manipulate dates using `lubridate` functions
- Extract information from text and dates for business insights
- Combine string and date operations for customer segmentation
- Create business-ready reports from raw data

## Instructions

- Complete all tasks in this notebook
- Write your code in the designated TODO sections
- Use the pipe operator (`%>%`) wherever possible
- Add comments explaining your logic
- Run all cells to verify your code works
- Answer all reflection questions

## Datasets

You will work with three CSV files:
- `customer_feedback.csv` - Customer reviews with messy text
- `transaction_log.csv` - Transaction records with dates
- `product_catalog.csv` - Product descriptions needing standardization

---

## Part 1: Data Import and Initial Exploration

**Business Context:** Before cleaning data, you must understand its structure and quality issues.

**Your Tasks:**
1. Load required packages (`tidyverse` and `lubridate`)
2. Import all three CSV files from the `data/` directory
3. Examine the structure and identify data quality issues
4. Display sample rows to understand the data

In [61]:
# Task 1.1: Load Required Packages
# TODO: Load tidyverse (includes stringr)
library(tidyverse)

# TODO: Load lubridate
library(lubridate)

cat("✅ Packages loaded successfully!\n")

✅ Packages loaded successfully!


In [62]:
# Task 1.2: Import Datasets
# TODO: Import customer_feedback.csv into a variable called 'feedback'
feedback <- read.csv("customer_feedback.csv")

# TODO: Import transaction_log.csv into a variable called 'transactions'
transactions <- read.csv("transaction_log.csv")

# TODO: Import product_catalog.csv into a variable called 'products'
products <- read.csv("product_catalog.csv")

cat("✅ Data imported successfully!\n")
cat("Feedback rows:", nrow(feedback), "\n")
cat("Transaction rows:", nrow(transactions), "\n")
cat("Product rows:", nrow(products), "\n")

✅ Data imported successfully!
Feedback rows: 100 
Transaction rows: 150 
Product rows: 75 


In [63]:
# Task 1.3: Initial Data Exploration

cat("=== CUSTOMER FEEDBACK DATA ===\n")
# TODO: Display structure of feedback using str()
str(feedback)

# TODO: Display first 5 rows of feedback
head(feedback, 5)

cat("\n=== TRANSACTION DATA ===\n")
# TODO: Display structure of transactions
str(transactions)

# TODO: Display first 5 rows of transactions
head(transactions, 5)

cat("\n=== PRODUCT CATALOG DATA ===\n")
# TODO: Display structure of products
str(products)

# TODO: Display first 5 rows of products
head(products, 5)


=== CUSTOMER FEEDBACK DATA ===
'data.frame':	100 obs. of  5 variables:
 $ FeedbackID   : int  1 2 3 4 5 6 7 8 9 10 ...
 $ CustomerID   : int  12 40 34 1 47 13 13 37 49 23 ...
 $ Feedback_Text: chr  "Highly recommend this item" "   Excellent service   " "Poor quality control" "average product, nothing special" ...
 $ Contact_Info : chr  "bob.wilson@test.org" "555-123-4567" "jane_smith@company.com" "jane_smith@company.com" ...
 $ Feedback_Date: chr  "2024-02-23" "2024-01-21" "2023-09-02" "2023-08-21" ...


FeedbackID CustomerID Feedback_Text                    Contact_Info          
1 1          12         Highly recommend this item       bob.wilson@test.org   
2 2          40            Excellent service             555-123-4567          
3 3          34         Poor quality control             jane_smith@company.com
4 4           1         average product, nothing special jane_smith@company.com
5 5          47         AMAZING customer support!!!      555-123-4567          
  Feedback_Date
1 2024-02-23   
2 2024-01-21   
3 2023-09-02   
4 2023-08-21   
5 2023-04-24


=== TRANSACTION DATA ===
'data.frame':	150 obs. of  5 variables:
 $ LogID               : int  1 2 3 4 5 6 7 8 9 10 ...
 $ CustomerID          : int  26 21 12 6 32 27 31 30 31 13 ...
 $ Transaction_DateTime: chr  "4/5/24 14:30" "3/15/24 14:30" "3/15/24 14:30" "3/20/24 9:15" ...
 $ Amount              : num  277 175 252 215 269 ...
 $ Status              : chr  "Pending" "Pending" "Pending" "Pending" ...


LogID CustomerID Transaction_DateTime Amount Status   
1 1     26         4/5/24 14:30         277.22 Pending  
2 2     21         3/15/24 14:30        175.16 Pending  
3 3     12         3/15/24 14:30        251.71 Pending  
4 4      6         3/20/24 9:15         214.98 Pending  
5 5     32         3/20/24 9:15         268.91 Completed


=== PRODUCT CATALOG DATA ===
'data.frame':	75 obs. of  5 variables:
 $ ProductID          : int  1 2 3 4 5 6 7 8 9 10 ...
 $ Product_Description: chr  "Apple iPhone 14 Pro - 128GB - Space Black" "samsung galaxy s23 ultra 256gb" "Apple iPhone 14 Pro - 128GB - Space Black" "Apple iPhone 14 Pro - 128GB - Space Black" ...
 $ Category           : chr  "TV" "TV" "Audio" "Shoes" ...
 $ Price              : num  964 1817 853 649 586 ...
 $ In_Stock           : chr  "Limited" "Yes" "Yes" "Yes" ...


ProductID Product_Description                       Category    Price  
1 1         Apple iPhone 14 Pro - 128GB - Space Black TV           963.53
2 2         samsung galaxy s23 ultra 256gb            TV          1817.44
3 3         Apple iPhone 14 Pro - 128GB - Space Black Audio        852.79
4 4         Apple iPhone 14 Pro - 128GB - Space Black Shoes        648.58
5 5         samsung galaxy s23 ultra 256gb            Electronics  586.35
  In_Stock
1 Limited 
2 Yes     
3 Yes     
4 Yes     
5 Limited

## Part 2: String Cleaning and Standardization

**Business Context:** Product names and feedback text often have inconsistent formatting that prevents accurate analysis.

**Your Tasks:**
1. Clean product names (remove extra spaces, standardize case)
2. Standardize product categories
3. Clean customer feedback text
4. Extract customer names from feedback

**Key Functions:** `str_trim()`, `str_squish()`, `str_to_lower()`, `str_to_upper()`, `str_to_title()`

In [64]:
# Task 2.1: Clean Product Names
# TODO: Create a new column 'product_name_clean' that:
#   - Removes leading/trailing whitespace using str_trim()
#   - Converts to Title Case using str_to_title()

products_clean <- products %>%
  mutate(
    product_name_clean = str_to_title(str_trim(Product_Description))
    
  )

# Display before and after
cat("Product Name Cleaning Results:\n")
products_clean %>%
  select(Product_Description, product_name_clean) %>%
  head(10) %>%
  print()

Product Name Cleaning Results:
                         Product_Description
1  Apple iPhone 14 Pro - 128GB - Space Black
2             samsung galaxy s23 ultra 256gb
3  Apple iPhone 14 Pro - 128GB - Space Black
4  Apple iPhone 14 Pro - 128GB - Space Black
5             samsung galaxy s23 ultra 256gb
6  Apple iPhone 14 Pro - 128GB - Space Black
7   DELL XPS 13 Laptop - Intel i7 - 16GB RAM
8         hp envy printer - wireless - color
9   Nike Air Max 270 - Size 10 - Black/White
10         LG 55" 4K Smart TV - OLED Display
                          product_name_clean
1  Apple Iphone 14 Pro - 128gb - Space Black
2             Samsung Galaxy S23 Ultra 256gb
3  Apple Iphone 14 Pro - 128gb - Space Black
4  Apple Iphone 14 Pro - 128gb - Space Black
5             Samsung Galaxy S23 Ultra 256gb
6  Apple Iphone 14 Pro - 128gb - Space Black
7   Dell Xps 13 Laptop - Intel I7 - 16gb Ram
8         Hp Envy Printer - Wireless - Color
9   Nike Air Max 270 - Size 10 - Black/White
10         Lg 55" 4k Sma

In [65]:
# Task 2.2: Standardize Product Categories
# TODO: Create a new column 'category_clean' that:
#   - Converts category to Title Case
#   - Removes any extra whitespace

products_clean <- products_clean %>%
  mutate(
    category_clean = str_to_title(str_trim(Category))
  )

# Show unique categories before and after
cat("Original categories:\n")
print(unique(products$Category))

cat("\nCleaned categories:\n")
print(unique(products_clean$category_clean))

Original categories:
[1] "TV"          "Audio"       "Shoes"       "Electronics" "Computers"  

Cleaned categories:
[1] "Tv"          "Audio"       "Shoes"       "Electronics" "Computers"  


In [66]:
# Task 2.3: Clean Customer Feedback Text
# TODO: Create a new column 'feedback_clean' that:
#   - Converts text to lowercase using str_to_lower()
#   - Removes extra whitespace using str_squish()

feedback_clean <- feedback %>%
  mutate(
    feedback_clean = str_squish(str_to_lower(Feedback_Text))
  )

# Display sample
cat("Feedback Cleaning Sample:\n")
feedback_clean %>%
    select(Feedback_Text, feedback_clean) %>%
  head(5) %>%
  print()

Feedback Cleaning Sample:
                     Feedback_Text                   feedback_clean
1       Highly recommend this item       highly recommend this item
2             Excellent service                   excellent service
3             Poor quality control             poor quality control
4 average product, nothing special average product, nothing special
5      AMAZING customer support!!!      amazing customer support!!!


## Part 3: Pattern Detection and Extraction

**Business Context:** Identifying products with specific features and extracting specifications helps with inventory management and marketing.

**Your Tasks:**
1. Identify products with specific keywords (wireless, premium, gaming)
2. Extract numerical specifications from product names
3. Detect sentiment words in customer feedback
4. Extract email addresses from feedback

**Key Functions:** `str_detect()`, `str_extract()`, `str_count()`

In [ ]:
# Task 3.1: Detect Product Features
# TODO: Create three new columns:
#   - is_wireless: TRUE if product name contains "wireless" (case-insensitive)
#   - is_premium: TRUE if product name contains "pro", "premium", or "deluxe"
#   - is_gaming: TRUE if product name contains "gaming" or "gamer"
# Hint: Use str_detect() with str_to_lower() for case-insensitive matching
# Hint: Use | (pipe) in regex for OR conditions

products_clean <- products_clean %>%
  mutate(
    is_wireless = str_detect(str_to_lower(product_name_clean), "wireless"),
    is_premium = str_detect(str_to_lower(product_name_clean), "pro|premium|deluxe"),
    is_gaming = str_detect(str_to_lower(product_name_clean), "gaming|gamer")
  )

# Display results
cat("Product Feature Detection:\n")
products_clean %>%
  select(product_name_clean, is_wireless, is_premium, is_gaming) %>%
  head(10) %>%
  print()

# Summary statistics
cat("\nFeature Summary:\n")
cat("Wireless products:", sum(products_clean$is_wireless), "\n")
cat("Premium products:", sum(products_clean$is_premium), "\n")
cat("Gaming products:", sum(products_clean$is_gaming), "\n")

Product Feature Detection:
                          product_name_clean is_wireless is_premium is_gaming
1  Apple Iphone 14 Pro - 128gb - Space Black       FALSE       TRUE     FALSE
2             Samsung Galaxy S23 Ultra 256gb       FALSE      FALSE     FALSE
3  Apple Iphone 14 Pro - 128gb - Space Black       FALSE       TRUE     FALSE
4  Apple Iphone 14 Pro - 128gb - Space Black       FALSE       TRUE     FALSE
5             Samsung Galaxy S23 Ultra 256gb       FALSE      FALSE     FALSE
6  Apple Iphone 14 Pro - 128gb - Space Black       FALSE       TRUE     FALSE
7   Dell Xps 13 Laptop - Intel I7 - 16gb Ram       FALSE      FALSE     FALSE
8         Hp Envy Printer - Wireless - Color        TRUE      FALSE     FALSE
9   Nike Air Max 270 - Size 10 - Black/White       FALSE      FALSE     FALSE
10         Lg 55" 4k Smart Tv - Oled Display       FALSE      FALSE     FALSE

Feature Summary:
Wireless products: 17 
Premium products: 13 
Gaming products: 0 


In [68]:
# Task 3.2: Extract Product Specifications
# TODO: Create a new column 'size_number' that extracts the first number from product_name
# Hint: Use str_extract() with pattern "\\\\d+" to match one or more digits

products_clean <- products_clean %>%
  mutate(
    size_number = str_extract(Product_Description, "\\d+")
  )

# Display products with extracted sizes
cat("Extracted Product Specifications:\n")
products_clean %>%
  filter(!is.na(size_number)) %>%
  select(product_name_clean, size_number) %>%
  head(10) %>%
  print()

Extracted Product Specifications:
                          product_name_clean size_number
1  Apple Iphone 14 Pro - 128gb - Space Black          14
2             Samsung Galaxy S23 Ultra 256gb          23
3  Apple Iphone 14 Pro - 128gb - Space Black          14
4  Apple Iphone 14 Pro - 128gb - Space Black          14
5             Samsung Galaxy S23 Ultra 256gb          23
6  Apple Iphone 14 Pro - 128gb - Space Black          14
7   Dell Xps 13 Laptop - Intel I7 - 16gb Ram          13
8   Nike Air Max 270 - Size 10 - Black/White         270
9          Lg 55" 4k Smart Tv - Oled Display          55
10       Sony Wh-1000xm4 Wireless Headphones        1000


In [69]:
# Task 3.3: Simple Sentiment Analysis
# TODO: Create three new columns:
#   - positive_words: count of positive words ("great", "excellent", "love", "amazing")
#   - negative_words: count of negative words ("bad", "terrible", "hate", "awful")
#   - sentiment_score: positive_words - negative_words
# Hint: Use str_count() to count pattern occurrences

feedback_clean <- feedback_clean %>%
  mutate(
    positive_words = str_count(str_to_lower(feedback_clean), "great|excellent|love|amazing"),
    negative_words = str_count(str_to_lower(feedback_clean), "bad|terrible|hate|awful"),
    sentiment_score = positive_words - negative_words
  )

# Display sentiment analysis results
cat("Sentiment Analysis Results:\n")
feedback_clean %>%
  select(feedback_clean, positive_words, negative_words, sentiment_score) %>%
  head(10) %>%
  print()

# Summary
cat("\nOverall Sentiment Summary:\n")
cat("Average sentiment score:", mean(feedback_clean$sentiment_score), "\n")
cat("Positive reviews:", sum(feedback_clean$sentiment_score > 0), "\n")
cat("Negative reviews:", sum(feedback_clean$sentiment_score < 0), "\n")

Sentiment Analysis Results:
                     feedback_clean positive_words negative_words
1        highly recommend this item              0              0
2                 excellent service              1              0
3              poor quality control              0              0
4  average product, nothing special              0              0
5       amazing customer support!!!              1              0
6       amazing customer support!!!              1              0
7  average product, nothing special              0              0
8              good value for money              0              0
9        highly recommend this item              0              0
10       highly recommend this item              0              0
   sentiment_score
1                0
2                1
3                0
4                0
5                1
6                1
7                0
8                0
9                0
10               0

Overall Sentiment Summary:
Average s

## Part 4: Date Parsing and Component Extraction

**Business Context:** Transaction dates need to be parsed and analyzed to understand customer behavior patterns.

**Your Tasks:**
1. Parse transaction dates from text to Date objects
2. Extract date components (year, month, day, weekday)
3. Identify weekend vs weekday transactions
4. Extract quarter and month names

**Key Functions:** `ymd()`, `mdy()`, `dmy()`, `year()`, `month()`, `day()`, `wday()`, `quarter()`

In [70]:
# Task 4.1: Parse Transaction Dates
# TODO: Create a new column 'date_parsed' that parses the transaction_date column
# Hint: Check the format of transaction_date first, then use ymd(), mdy(), or dmy()

transactions_clean <- transactions %>%
  mutate(
    date_parsed = parse_date_time(Transaction_DateTime,
    orders = c("mdy HM", "dmy HMS","ymd HMS"))
    
  )

# Verify parsing worked
cat("Date Parsing Results:\n")
transactions_clean %>%
  select(Transaction_DateTime, date_parsed) %>%
  head(10) %>%
  print()

Date Parsing Results:
   Transaction_DateTime         date_parsed
1          4/5/24 14:30 2024-04-05 14:30:00
2         3/15/24 14:30 2024-03-15 14:30:00
3         3/15/24 14:30 2024-03-15 14:30:00
4          3/20/24 9:15 2024-03-20 09:15:00
5          3/20/24 9:15 2024-03-20 09:15:00
6          3/20/24 9:15 2024-03-20 09:15:00
7          3/20/24 9:15 2024-03-20 09:15:00
8         3/15/24 14:30 2024-03-15 14:30:00
9   25-03-2024 16:45:30 2024-03-25 16:45:30
10         4/5/24 14:30 2024-04-05 14:30:00


In [71]:
# Task 4.2: Extract Date Components
# TODO: Create the following new columns:
#   - trans_year: Extract year from date_parsed
#   - trans_month: Extract month number from date_parsed
#   - trans_month_name: Extract month name (use label=TRUE, abbr=FALSE)
#   - trans_day: Extract day of month from date_parsed
#   - trans_weekday: Extract weekday name (use label=TRUE, abbr=FALSE)
#   - trans_quarter: Extract quarter from date_parsed

transactions_clean <- transactions_clean %>%
  mutate(
    trans_year = year(date_parsed), 
    trans_month = month(date_parsed),
    trans_month_name = month(date_parsed, label = TRUE, abbr = FALSE), 
    trans_day = day(date_parsed),
    trans_weekday = wday(date_parsed, label = TRUE, abbr = FALSE),
    trans_quarter = quarter(date_parsed)
  )

# Display results
cat("Date Component Extraction:\n")
transactions_clean %>%
  select(date_parsed, trans_month_name, trans_weekday, trans_quarter) %>%
  head(10) %>%
  print()

Date Component Extraction:
           date_parsed trans_month_name trans_weekday trans_quarter
1  2024-04-05 14:30:00            April        Friday             2
2  2024-03-15 14:30:00            March        Friday             1
3  2024-03-15 14:30:00            March        Friday             1
4  2024-03-20 09:15:00            March     Wednesday             1
5  2024-03-20 09:15:00            March     Wednesday             1
6  2024-03-20 09:15:00            March     Wednesday             1
7  2024-03-20 09:15:00            March     Wednesday             1
8  2024-03-15 14:30:00            March        Friday             1
9  2024-03-25 16:45:30            March        Monday             1
10 2024-04-05 14:30:00            April        Friday             2


In [72]:
# Task 4.3: Identify Weekend Transactions
# TODO: Create a new column 'is_weekend' that is TRUE if the transaction was on Saturday or Sunday
# Hint: Use wday() which returns 1 for Sunday and 7 for Saturday
# Hint: Use %in% c(1, 7) to check if day is weekend

transactions_clean <- transactions_clean %>%
  mutate(
    is_weekend = wday(date_parsed) %in% c(1, 7)
    
  )

# Summary
cat("Weekend vs Weekday Transactions:\n")
table(transactions_clean$is_weekend) %>% print()

cat("\nPercentage of weekend transactions:",
    round(sum(transactions_clean$is_weekend) / nrow(transactions_clean) * 100, 1), "%\n")

Weekend vs Weekday Transactions:

FALSE 
  150 

Percentage of weekend transactions: 0 %


## Part 5: Date Calculations and Customer Recency Analysis

**Business Context:** Understanding how recently customers transacted helps identify at-risk customers for re-engagement campaigns.

**Your Tasks:**
1. Calculate days since each transaction
2. Categorize customers by recency (Recent, Moderate, Old)
3. Identify customers who haven't transacted in 90+ days
4. Calculate average days between transactions per customer

**Key Functions:** `today()`, date arithmetic, `case_when()`

In [73]:
# Task 5.1: Calculate Days Since Transaction
# TODO: Create a new column 'days_since' that calculates days from date_parsed to today()
# Hint: Use as.numeric(today() - date_parsed)

transactions_clean <- transactions_clean %>%
  mutate(
    days_since = as.numeric(today() - as_date(date_parsed))
    
  )

# Display results
cat("Days Since Transaction:\n")
transactions_clean %>%
  select(CustomerID, date_parsed, days_since) %>%
  arrange(desc(days_since)) %>%
  head(10) %>%
  print()

Days Since Transaction:
   CustomerID         date_parsed days_since
1          21 2024-03-15 14:30:00        584
2          12 2024-03-15 14:30:00        584
3          30 2024-03-15 14:30:00        584
4          45 2024-03-15 14:30:00        584
5           2 2024-03-15 14:30:00        584
6          18 2024-03-15 14:30:00        584
7          34 2024-03-15 14:30:00        584
8          48 2024-03-15 14:30:00        584
9          28 2024-03-15 14:30:00        584
10         30 2024-03-15 14:30:00        584


In [74]:
# Task 5.2: Categorize by Recency
# TODO: Create a new column 'recency_category' using case_when():
#   - "Recent" if days_since <= 30
#   - "Moderate" if days_since <= 90
#   - "At Risk" if days_since > 90

transactions_clean <- transactions_clean %>%
  mutate(
    recency_category = case_when(
      days_since <= 30 ~ "Recent",
      days_since <= 90 ~ "Moderate",
      days_since > 90 ~ "At Risk"
    )
  )

# Display distribution
cat("Recency Category Distribution:\n")
table(transactions_clean$recency_category) %>% print()

# Show at-risk customers
cat("\nAt-Risk Customers (>90 days):\n")
transactions_clean %>%
  filter(recency_category == "At Risk") %>%
  select(CustomerID, date_parsed, days_since) %>%
  arrange(desc(days_since)) %>%
  print()

Recency Category Distribution:

At Risk 
    150 

At-Risk Customers (>90 days):
    CustomerID         date_parsed days_since
1           21 2024-03-15 14:30:00        584
2           12 2024-03-15 14:30:00        584
3           30 2024-03-15 14:30:00        584
4           45 2024-03-15 14:30:00        584
5            2 2024-03-15 14:30:00        584
6           18 2024-03-15 14:30:00        584
7           34 2024-03-15 14:30:00        584
8           48 2024-03-15 14:30:00        584
9           28 2024-03-15 14:30:00        584
10          30 2024-03-15 14:30:00        584
11          33 2024-03-15 14:30:00        584
12          11 2024-03-15 14:30:00        584
13          36 2024-03-15 14:30:00        584
14           6 2024-03-15 14:30:00        584
15           8 2024-03-15 14:30:00        584
16          38 2024-03-15 14:30:00        584
17          49 2024-03-15 14:30:00        584
18          28 2024-03-15 14:30:00        584
19          44 2024-03-15 14:30:00        584

## Part 6: Combined String and Date Operations

**Business Context:** Create personalized customer outreach messages based on purchase recency.

**Your Tasks:**
1. Extract first names from customer names
2. Create personalized messages based on recency
3. Analyze transaction patterns by weekday
4. Identify best customers (recent + high value)

**Key Functions:** Combine `str_extract()`, date calculations, `case_when()`, `group_by()`, `summarize()`

In [75]:
# Task 6.1: Extract First Names and Create Personalized Messages
# TODO: Create two new columns:
#   - first_name: Extract first name from customer_name (everything before first space)
#   - personalized_message: Create message based on recency_category
#     * Recent: "Hi [name]! Thanks for your recent purchase!"
#     * Moderate: "Hi [name], we miss you! Check out our new products."
#     * At Risk: "Hi [name], it's been a while! Here's a special offer for you."
# Hint: Use str_extract() with pattern "^\\\\w+" for first name
# Hint: Use paste() to combine strings in case_when()

customer_outreach <- transactions_clean %>%
  mutate(
    customer_name = paste("Customer", CustomerID),
    first_name = str_extract(customer_name, "^\\w+"),
        personalized_message = case_when(
      recency_category == "Recent" ~ paste("Hi", first_name, "! Thanks for your recent purchase!"),
      recency_category == "Moderate" ~ paste("Hi", first_name, ", we miss you! Check out our new products."),
      recency_category == "At Risk" ~ paste("Hi", first_name, ", it's been a while! Here's a special offer for you.")
    )
  )

# Display personalized messages
cat("Personalized Customer Messages:\n")
customer_outreach %>%
  select(customer_name, first_name, days_since, personalized_message) %>%
  head(10) %>%
  print()

Personalized Customer Messages:
   customer_name first_name days_since
1    Customer 26   Customer        563
2    Customer 21   Customer        584
3    Customer 12   Customer        584
4     Customer 6   Customer        579
5    Customer 32   Customer        579
6    Customer 27   Customer        579
7    Customer 31   Customer        579
8    Customer 30   Customer        584
9    Customer 31   Customer        574
10   Customer 13   Customer        563
                                               personalized_message
1  Hi Customer , it's been a while! Here's a special offer for you.
2  Hi Customer , it's been a while! Here's a special offer for you.
3  Hi Customer , it's been a while! Here's a special offer for you.
4  Hi Customer , it's been a while! Here's a special offer for you.
5  Hi Customer , it's been a while! Here's a special offer for you.
6  Hi Customer , it's been a while! Here's a special offer for you.
7  Hi Customer , it's been a while! Here's a special offer for 

In [76]:
# Task 6.2: Analyze Transaction Patterns by Weekday
# TODO: Group by trans_weekday and calculate:
#   - transaction_count: number of transactions
#   - total_amount: sum of amount (if available)
#   - avg_amount: average amount per transaction
# TODO: Arrange by transaction_count descending

weekday_patterns <- transactions_clean %>%
  group_by(trans_weekday) %>%
  summarize(
    transaction_count = n(),
    total_amount = sum(Amount, na.rm = TRUE),
    avg_amount = mean(Amount, na.rm = TRUE)
  ) %>%
  arrange(desc(transaction_count))

# Display results
cat("Transaction Patterns by Weekday:\n")
print(weekday_patterns)

# Identify busiest day
busiest_day <- weekday_patterns$trans_weekday[1]
cat("\n🔥 Busiest day:", as.character(busiest_day), "\n")

Transaction Patterns by Weekday:


# A tibble: 3 × 4
  trans_weekday transaction_count total_amount avg_amount
  <ord>                     <int>        <dbl>      <dbl>
1 Monday                       61       15367.       252.
2 Friday                       55       14789.       269.
3 Wednesday                    34        7578.       223.

🔥 Busiest day: Monday 


In [77]:
# Task 6.3: Monthly Transaction Analysis
# TODO: Group by trans_month_name and calculate:
#   - transaction_count
#   - unique_customers: use n_distinct(customer_name)
# TODO: Arrange by trans_month (to show chronological order)

monthly_patterns <- transactions_clean %>%
  group_by(trans_month_name, trans_month)  %>%
  summarize( 
    transaction_count = n(),
    unique_customers = n_distinct(CustomerID),
  .groups = "drop"
  ) %>%
  arrange(trans_month)
# Display results
cat("Monthly Transaction Patterns:\n")
print(monthly_patterns)

Monthly Transaction Patterns:
# A tibble: 2 × 4
  trans_month_name trans_month transaction_count unique_customers
  <ord>                  <dbl>             <int>            <int>
1 March                      3                89               37
2 April                      4                61               34


## Part 7: Business Intelligence Summary

**Business Context:** Create an executive summary that combines all your analyses into actionable insights.

**Your Tasks:**
1. Calculate key metrics across all datasets
2. Identify top products and categories
3. Summarize customer sentiment
4. Provide data-driven recommendations

In [78]:
# Task 7.1: Create Business Intelligence Dashboard

cat("\n", rep("=", 60), "\n")
cat("         BUSINESS INTELLIGENCE SUMMARY\n")
cat(rep("=", 60), "\n\n")

# Product Analysis
cat("📦 PRODUCT ANALYSIS\n")
cat(rep("─", 30), "\n")
# TODO: Calculate and display:
#   - Total number of products
total_products <- nrow(products_clean)
#   - Number of wireless products
wireless_count <- sum(products_clean$is_wireless, na.rm = TRUE)
#   - Number of premium products
premium_count <- sum(products_clean$is_premium, na.rm = TRUE)
#   - Most common category
most_common_category <- names(sort(table(products_clean$Category), decreasing = TRUE))[1]

cat("Total Products:", total_products, "\n")
cat("Wireless Products:", wireless_count, "(", round(wireless_count/total_products*100, 1), "%)\n")
cat("Premium Products:", premium_count, "(", round(premium_count/total_products*100, 1), "%)\n")
cat("Most Common Category:", most_common_category, "\n")

# Customer Sentiment
cat("\n💬 CUSTOMER SENTIMENT\n")
cat(rep("─", 30), "\n")
# TODO: Calculate and display:
#   - Total feedback entries
total_feedback <- nrow(feedback_clean)
#   - Average sentiment score
avg_sentiment <- mean(feedback_clean$sentiment_score, na.rm = TRUE)
#   - Percentage of positive reviews
positive_reviews <- sum(feedback_clean$sentiment_score > 0, na.rm = TRUE)
pct_positive <- round(positive_reviews/total_feedback*100, 1)
#   - Percentage of negative reviews
negative_reviews <- sum(feedback_clean$sentiment_score < 0, na.rm = TRUE)
pct_negative <- round(negative_reviews/total_feedback*100, 1)

cat("Total Feedback Entries:", total_feedback, "\n")
cat("Average Sentiment Score:", avg_sentiment, "\n")
cat("Positive Reviews:", positive_reviews, "(", pct_positive, "%)\n")
cat("Negative Reviews:", negative_reviews, "(", pct_negative, "%)\n")

# Transaction Patterns
cat("\n📊 TRANSACTION PATTERNS\n")
cat(rep("─", 30), "\n")
# TODO: Calculate and display:
#   - Total transactions
total_transactions <- nrow(transactions_clean)
#   - Date range (earliest to latest)
earliest_date <- min(transactions_clean$date_parsed, na.rm = TRUE)
latest_date <- max(transactions_clean$date_parsed, na.rm = TRUE)
#   - Busiest weekday
busiest_weekday <- as.character(busiest_day)

#   - Weekend transaction percentage
weekend_pct <- sum(transactions_clean$is_weekend, na.rm = TRUE) / total_transactions * 100

cat("Total Transactions:", total_transactions, "\n")
cat("Date Range:", format(earliest_date, "%Y-%m-%d"), "to", format(latest_date, "%Y-%m-%d"), "\n")
cat("Busiest Weekday:", busiest_weekday, "\n")
cat("Weekend Transactions:", weekend_pct, "%\n")

# Customer Recency
cat("\n👥 CUSTOMER RECENCY\n")
cat(rep("─", 30), "\n")
# TODO: Calculate and display:
#   - Number of recent customers (< 30 days)
recent_customers <- sum(transactions_clean$days_since < 30, na.rm = TRUE)
#   - Number of at-risk customers (> 90 days)
at_risk_customers <- sum(transactions_clean$days_since > 90, na.rm = TRUE)
#   - Percentage needing re-engagement
pct_at_risk <- round(at_risk_customers/total_transactions*100, 1)

cat("Recent Customers (under 30 days):", recent_customers, "\n")
cat("At-Risk Customers (over 90 days):", at_risk_customers, "\n")
cat("Customers Needing Re-engagement:", pct_at_risk, "%\n")



 = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = 
         BUSINESS INTELLIGENCE SUMMARY
= = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = 

📦 PRODUCT ANALYSIS
─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ 


Total Products: 75 
Wireless Products: 17 ( 22.7 %)
Premium Products: 13 ( 17.3 %)
Most Common Category: Electronics 

💬 CUSTOMER SENTIMENT
─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ 
Total Feedback Entries: 100 
Average Sentiment Score: 0.18 
Positive Reviews: 30 ( 30 %)
Negative Reviews: 20 ( 20 %)

📊 TRANSACTION PATTERNS
─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ 
Total Transactions: 150 
Date Range: 2024-03-15 to 2024-04-05 
Busiest Weekday: Monday 
Weekend Transactions: 0 %

👥 CUSTOMER RECENCY
─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ 
Recent Customers (under 30 days): 0 
At-Risk Customers (over 90 days): 150 
Customers Needing Re-engagement: 100 %


add recommendations 

In [79]:
# Task 7.2: Identify Top Products by Category
# TODO: Group products by category_clean and count products in each
# TODO: Arrange by count descending
# TODO: Display top 5 categories

top_categories <- products_clean %>%
    group_by(Category) %>%
  summarise(product_count = n()) %>%
  arrange(desc(product_count)) %>%
  head(5)
  

cat("Top Product Categories:\n")
print(top_categories)

Top Product Categories:
# A tibble: 5 × 2
  Category    product_count
  <chr>               <int>
1 Electronics            21
2 Computers              15
3 Audio                  14
4 TV                     14
5 Shoes                  11


## Part 8: Reflection Questions

Answer the following questions based on your analysis. Write your answers in the markdown cells below.

### Question 8.1: Data Quality Impact

**How did cleaning the text data (removing spaces, standardizing case) improve your ability to analyze the data? Provide specific examples from your homework.**

Helps fix inconsistent formatting throughout the raw data using the str_trim() and str_to_title() functions to give consistent standardization for all product names. Without the cleaning step pattern matching would’ve failed and lead to missed insights. In Task 2.3 I standardized the feedback text using the code feedback_clean = str_squish(str_to_lower(Feedback_Text)) to convert all text lowercase and get rid of whitespace into the feedback_clean file. Then later in task 3.3 Sentiment analysis when using feedback_clean file using the str_count() on great, excellent, love, and amazing words. It allowed the function to count any of the words that were already put in lowercase in a previous step. It turned feedback with “AMAZING” to “amazing” and it allowed me to better analyze the feedback_clean file in sentiment analysis accuracy.  




### Question 8.2: Pattern Detection Value

**What business insights did you gain from detecting patterns in product names (wireless, premium, gaming)? How could a business use this information?**

Specific findings I found were 17 wireless products and 13 premium products but no gaming products. This insight reveals the strengths and gaps in purchased products and a business could use this insight to make marketing strategies. For example, create a bundling strategy for wireless products with items that can go together with gaming items with lower sales such as wireless headphones paired with an xbox. This would help tap into the gaming market paired with products that are already consistently being sold. In addition to this, customers who already have premium products could be sent recommendations to other high end products. This is due to the fact they’ve purchased higher end items and would be more likely to purchase more if receiving targeted promotions. 



### Question 8.3: Date Analysis Importance

**Why is analyzing transaction dates by weekday and month important for business operations? Provide at least three specific business applications.**



1. The busiest day of transactions was shown to take place on Monday with 61 transactions and slowest day on Wednesday with 34, this could affect the way a business would schedule staff. More staff would be required on mondays to deal with the more customer transactions taking place to ensure operations are taking place efficiently, and reduce staff on Wednesday to optimize labor cost by having a prediction on how many sales will take place on those days. 

2. March had 89 transactions which could require more sales events or product releases to use the momentum of sales to have better selling effectiveness. This could also require larger marketing budgets to get a better use on money spent on ads. On the other hand April had 61 transactions showing a drop in customers sales, for slower months like this a company could lower its marketing budget. Using discounts could help stimulate customers to help generate more sales during a slower month to help drive revenue. Instead of ads a company could use promotional emails to save money and help bring attention to customers about the discounts. 

3. Along with the High transactions in March, a business would need to make sure inventory is prepared and stocked before the high demand to make sure all transactions can be made. Then in the lower transactions in April could see a business reduce inventory to focus on clearing any remaining inventory from stocking in March and reducing any holding costs of unsold items. 


### Question 8.4: Customer Recency Strategy

**Based on your recency analysis, what specific actions would you recommend for customers in each category (Recent, Moderate, At Risk)? How would you prioritize these actions?**

All of the 150 customers are at risk and listed as at-risk customers with the last purchase taking place 563 days ago. This would require an aggressive business strategy for each category. Starting with customers that left positive reviews which are the most likely to make purchases again would be to launch targeted campaigns with personalized messaging. Next would be to address issues from negative reviews and send emails with discounts to encourage customers that have negative reviews to consider making future purchases. Long-term would be to keep customers re-engaged with automated messages once purchases are made to focus on customer retention. 




### Question 8.5: Sentiment Analysis Application

**How could the sentiment analysis you performed be used to improve products or customer service? What are the limitations of this simple sentiment analysis approach?**

Sentiment analysis could be used to improve customer services by tracking recurring negative reviews and address those issues such as quality control. Prioritizing products with negative feedback and combining with transaction data to better identify which products are generating negative reviews and improving the quality or supplier of those items. The limitation of sentiment analysis is it only counts specific words and can fail to give a full representation of how customers feel towards products. This could create limits to how quality control can be addressed.    



### Question 8.6: Real-World Application

**Describe a real business scenario where you would need to combine string manipulation and date analysis (like you did in this homework). What insights would you be trying to discover?**

An online retailer could require a full string manipulation and data analysis to help create re-engagement for customers who haven’t purchased in a while. Having a data analysis could reveal an urgency to prioritize an outreach to customers based on how long since a purchase has been made. This would be more valuable to an online retailer opposed to an in-person location due to the fact that online customers are easier to lose interest because of no physical store customers are easier to turn away so customer retention is that much more valuable. An online retailer could create automated personalized messages to customers based on risk such as what was completed in the homework in Task 8.6. 




## Summary and Submission

### What You've Accomplished

In this homework, you've successfully:
- ✅ Cleaned and standardized messy text data using `stringr` functions
- ✅ Detected patterns and extracted information from text
- ✅ Parsed dates and extracted temporal components using `lubridate`
- ✅ Calculated customer recency for segmentation
- ✅ Analyzed transaction patterns by time periods
- ✅ Combined string and date operations for business insights
- ✅ Created personalized customer communications
- ✅ Generated executive-ready business intelligence summaries

### Key Skills Mastered

**String Manipulation:**
- `str_trim()`, `str_squish()` - Whitespace handling
- `str_to_lower()`, `str_to_upper()`, `str_to_title()` - Case conversion
- `str_detect()` - Pattern detection
- `str_extract()` - Information extraction
- `str_count()` - Pattern counting

**Date/Time Operations:**
- `ymd()`, `mdy()`, `dmy()` - Date parsing
- `year()`, `month()`, `day()`, `wday()` - Component extraction
- `quarter()` - Period extraction
- `today()` - Current date
- Date arithmetic - Calculating differences

**Business Applications:**
- Data cleaning and standardization
- Customer segmentation by recency
- Sentiment analysis
- Pattern identification
- Temporal trend analysis
- Personalized communication

### Submission Checklist

Before submitting, ensure you have:
- [ ] Entered your name, student ID, and date at the top
- [ ] Completed all code tasks (Parts 1-7)
- [ ] Run all cells successfully without errors
- [ ] Answered all reflection questions (Part 8)
- [ ] Used proper commenting in your code
- [ ] Used the pipe operator (`%>%`) where appropriate
- [ ] Verified your results make business sense
- [ ] Checked for any remaining TODO comments

### Grading Criteria

Your homework will be evaluated on:
- **Code Correctness (40%)**: All tasks completed correctly
- **Code Quality (20%)**: Clean, well-commented, efficient code
- **Business Understanding (20%)**: Demonstrates understanding of business applications
- **Reflection Questions (15%)**: Thoughtful, complete answers
- **Presentation (5%)**: Professional formatting and organization

### Next Steps

In Lesson 8, you'll learn:
- Advanced data wrangling with complex pipelines
- Sophisticated conditional logic with `case_when()`
- Data validation and quality checks
- Creating reproducible analysis workflows
- Professional best practices for business analytics

**Great work on completing this assignment! 🎉**